R/extract.R

```diff
@@ -0,0 +1,174 @@
+#' ```r
+#'    workflow_fit %>% extract_fit_engine()  %>% predict(new_data)
+#'    # or
+#'    workflow_fit %>% extract_fit_parsnip() %>% predict(new_data)
+#' ```
+#' @return
+#' The extracted value from the object, `x`, as described in the description
+#' section.
+#'
+#' @name extract-workflow
+#' @examples
+#' library(parsnip)
+#' library(recipes)
+#' library(magrittr)
+#'
+#' model <- linear_reg() %>%
+#'   set_engine("lm")
+#'
+#' recipe <- recipe(mpg ~ cyl + disp, mtcars) %>%
+#'   step_log(disp)
+#'
+#' base_wf <- workflow() %>%
+#'   add_model(model)
+#'
+#' recipe_wf <- add_recipe(base_wf, recipe)
+#' formula_wf <- add_formula(base_wf, mpg ~ cyl + log(disp))
+#' variable_wf <- add_variables(base_wf, mpg, c(cyl, disp))
+#'
+#' fit_recipe_wf <- fit(recipe_wf, mtcars)
+#' fit_formula_wf <- fit(formula_wf, mtcars)
+#'
+#' # The preprocessor is a recipes, formula, or a list holding the
+#' # tidyselect expressions identifying the outcomes/predictors
+#' extract_preprocessor(recipe_wf)
+#' extract_preprocessor(formula_wf)
+#' extract_preprocessor(variable_wf)
+#'
+#' # The `spec` is the parsnip spec before it has been fit.
+#' # The `fit` is the fit parsnip model.
+#' extract_spec_parsnip(fit_formula_wf)
+#' extract_fit_parsnip(fit_formula_wf)
+#' extract_fit_engine(fit_formula_wf)
+#'
+#' # The mold is returned from `hardhat::mold()`, and contains the
+#' # predictors, outcomes, and information about the preprocessing
+#' # for use on new data at `predict()` time.
+#' extract_mold(fit_recipe_wf)
+#'
+#' # A useful shortcut is to extract the fitted recipe from the workflow
+#' extract_recipe(fit_recipe_wf)
+#'
+#' # That is identical to
+#' identical(
+#'   extract_mold(fit_recipe_wf)$blueprint$recipe,
+#'   extract_recipe(fit_recipe_wf)
+#' )
+NULL
+
+#' @export
+#' @rdname extract-workflow
+extract_spec_parsnip.workflow <- function(x, ...) {
+  if (has_spec(x)) {
+    return(x$fit$actions$model$spec)
+  }
+  abort("The workflow does not have a model spec.")
+}
+
+#' @export
+#' @rdname extract-workflow
+extract_recipe.workflow <- function(x, estimated = TRUE, ...) {
+  if (!has_preprocessor_recipe(x)) {
+    abort("The workflow must have a recipe preprocessor.")
+  }
+  if (estimated) {
+    # Gracefully fails if not yet fitted
+    mold <- extract_mold(x)
+    res <- mold$blueprint$recipe
+  } else {
+    res <- x$pre$actions$recipe$recipe
+  }
+  res
+}
+
+#' @export
+#' @rdname extract-workflow
+extract_fit_parsnip.workflow <- function(x, ...) {
+  if (has_fit(x)) {
+    return(x$fit$fit)
+  }
+  abort("The workflow does not have a model fit. Have you called `fit()` yet?")
+}
+
+#' @export
+#' @rdname extract-workflow
+extract_fit_engine.workflow <- function(x, ...) {
+  if (has_fit(x)) {
+    return(extract_fit_parsnip(x)$fit)
+  }
+  abort("The workflow does not have a model fit. Have you called `fit()` yet?")
+}
+
+#' @export
+#' @rdname extract-workflow
+extract_mold.workflow <- function(x, ...) {
+  validate_is_workflow(x)
```

Comment:
Don't need this since we are doing S3 dispatch now. It had to be a workflow

Reply:
You mean the `@export`?

Reply:
Sorry, I'm talking about `validate_is_workflow(x)`. That is what we don't need now.

Since this works off S3 dispatch, and this is a method for a workflow, then `x` must be a workflow so we don't need to validate that it is one